metaflow/decorators.py

```diff
@@ -239,6 +239,12 @@ def package_init(self, flow, step_name, environment):
         """
         pass
 
+    def add_to_package(self):
+        """
+        Called to add custom packages needed for a decorator
```

Comment:
Be more descriptive about what this should return. It seems its a tuple of (path, name) although see my comment about the name.

Reply:
Quick follow up, what problem does this PR help to solve? Is it an alternative to system pip install?

Reply:
This is not an alternative to pip install. 

### Context 

This PR introduces a decorator hook that allows the decorator to include files in the metaflow code package. As decorators plug into different abstractions of the Metaflow stack, we may require custom packaging based on the type of decorator we are implementing. This is where that hook becomes essential. 

### Example 

In #815  the `@card` decorator introduces the concept of custom card modules (`metaflow_cards`). These modules are external packages that may be present on your local python path; Including those for remote execution requires those modules to be a part of the Metaflow code package ( or installed externally via `@conda`) 

Hence For this decorator, We detect and bundle along those modules [with the metaflow code package](https://github.com/valayDave/metaflow/blob/f9d88cd8c13c48e691585d8b3e642d9c863b6b51/metaflow/plugins/cards/card_decorator.py#L60).

Reply:
> This is not an alternative to pip install.
>In #815 the @card decorator introduces the concept of custom card modules (metaflow_cards). These modules are external packages that may be present on your local python path; Including those for remote execution requires those modules to be a part of the Metaflow code package ( or installed externally via @conda)

how would one specify a custom package here that is not installable via `conda`?

Reply:
I am a little confused to how to answer this so I am decompressing what I wrote earlier. Apologies if this seems like an information dump or if I don't cover what you may be seeking for.

Just for clarification: As this is a decorator hook, these custom packages may be present to fulfill the requirements of the decorator. So they may not be a part of "userland" i.e the users don't need to explicitly specify the packages in the decorator invocation but rather the decorator automatically seeks them and includes them with metaflow code package. 

These packages may have been pip installed into the local python environment or present in PYTHONPATH or lie somewhere in the filesystem. Continuing the example of `@card` decorator, `@card` decorator seeks for `metaflow_cards` modules present in the local python path / pip packages.  The `metaflow_cards` module in the current context is a plugin-like module which behaves like a namespace package.  As a part of the MF run lifecycle we seek these modules from the current python path / pip packages and include them in the code package. We can even install them via `conda` if they are published on a `conda` channel. 

The main purpose of this hook is to include files to package before run starts ( code files that are collected through the `MetaflowPackage` class ).  

### Why do we use this hook 
The @card decorator uses other types of files except for python files such as `html` , `css`, and `js` files to create a HTML that visualize the results of a `Task`; This requires that we seek filetypes that are generally not included in the metaflow code package. Hence this hook helps seek and include them as a part of the code package. Even the custom `metaflow_cards` modules that we install can have other types of files like `html` , `css`, `js`, etc which get included by default because of this decorator hook. 

### Example of `metaflow_cards` Namespace package. 

As `metaflow_cards` is a namespace package; we can install it's subpackages using `pip install metaflow-cards-container-monitoring`; Once these packages are installed in the local python path or in pip library, they will be seeked by the card decorator and then included in the metaflow code package. 

###  Gotacha's with directly seeking Installed Modules
As we are directly seeking modules and including them in the code package, the dependencies needed by these packages may not be covered ; This is something which we don't seek (at the moment)( haven't thought deeply upon this) . 

Hope this addresses what you were seeking for ? :)

Reply:
yes, thank you for the details! My initial understanding of this change was incorrect - I had thought this would allow anyone to specify a list of packages required by the decorator and would attempt to `pip install` these list of packages. After talking to @romain-intel I have more context on this change. IIUC, this change includes the files required by the decorator in the code package. The files themselves could have been installed through one of several ways: `pip`, `conda` or simply copying the files on the FS.

While this doesn't solve the larger custom package and dependency management problem that we have, I think this work well for a certain set of use cases that we have internally.

Thanks for the PR and the explanation!